File: sdk/go/client_test.go
package gatewise

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"testing"
	"time"
)

const (
	testBaseURL = "http://gateway.test"
	testAPIKey  = "gatewise_sk_test_123"
)

// captured records the details of the request the handler received so a test can assert
// on the exact verb, path, query, headers and body the client produced.
type captured struct {
	method string
	path   string
	query  url.Values
	header http.Header
	body   []byte
}

// recordHandler records the incoming request into cap and replies with the given status
// and JSON body.
func recordHandler(cap *captured, status int, respBody string) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		cap.method = r.Method
		cap.path = r.URL.Path
		cap.query = r.URL.Query()
		cap.header = r.Header.Clone()
		cap.body, _ = io.ReadAll(r.Body)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(status)
		_, _ = io.WriteString(w, respBody)
	}
}

// newTestClient spins up an httptest server with the given handler and returns a client
// pointed at it, plus the server (which the test should Close).
func newTestClient(t *testing.T, handler http.Handler) (*Client, *httptest.Server) {
	t.Helper()
	srv := httptest.NewServer(handler)
	client, err := New(WithBaseURL(srv.URL), WithAPIKey(testAPIKey), WithHTTPClient(srv.Client()))
	if err != nil {
		srv.Close()
		t.Fatalf("New: %v", err)
	}
	return client, srv
}

// roundTripperFunc adapts a function to an http.RoundTripper for transport-level tests.
type roundTripperFunc func(*http.Request) (*http.Response, error)

func (f roundTripperFunc) RoundTrip(r *http.Request) (*http.Response, error) { return f(r) }

// -- configuration / env --------------------------------------------------------------

func TestEnvVarDefaults(t *testing.T) {
	t.Setenv("GATEWISE_BASE_URL", "http://env-host:9000")
	t.Setenv("GATEWISE_API_KEY", "gatewise_sk_env")
	client, err := New()
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	if client.baseURL != "http://env-host:9000" {
		t.Errorf("baseURL = %q, want http://env-host:9000", client.baseURL)
	}
	if client.apiKey != "gatewise_sk_env" {
		t.Errorf("apiKey = %q, want gatewise_sk_env", client.apiKey)
	}
}

func TestExplicitOptionsBeatEnv(t *testing.T) {
	t.Setenv("GATEWISE_BASE_URL", "http://env-host:9000")
	t.Setenv("GATEWISE_API_KEY", "gatewise_sk_env")
	client, err := New(WithBaseURL("http://explicit:1234"), WithAPIKey("gatewise_sk_explicit"))
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	if client.baseURL != "http://explicit:1234" {
		t.Errorf("baseURL = %q, want the explicit value", client.baseURL)
	}
	if client.apiKey != "gatewise_sk_explicit" {
		t.Errorf("apiKey = %q, want the explicit value", client.apiKey)
	}
}

func TestDefaultBaseURL(t *testing.T) {
	t.Setenv("GATEWISE_BASE_URL", "")
	client, err := New(WithAPIKey("gatewise_sk_x"))
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	if client.baseURL != "http://localhost:8080" {
		t.Errorf("baseURL = %q, want the localhost default", client.baseURL)
	}
}

func TestMissingAPIKeyIsAnError(t *testing.T) {
	t.Setenv("GATEWISE_API_KEY", "")
	if _, err := New(WithBaseURL(testBaseURL)); err == nil {
		t.Fatal("expected an error when no API key is configured, got nil")
	}
}

func TestTrailingSlashStripped(t *testing.T) {
	client, err := New(WithBaseURL(testBaseURL+"/"), WithAPIKey(testAPIKey))
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	if got := client.url("usage"); got != testBaseURL+"/v1/usage" {
		t.Errorf("url(usage) = %q, want %s/v1/usage", got, testBaseURL)
	}
}

// -- auth header + URL joining --------------------------------------------------------

func TestAuthHeaderAndContentTypeSent(t *testing.T) {
	var cap captured
	client, srv := newTestClient(t, recordHandler(&cap, 200, `{"choices":[]}`))
	defer srv.Close()

	_, err := client.Chat.Completions.Create(context.Background(), ChatCompletionRequest{
		Model:    "smart",
		Messages: []Message{{Role: "user", Content: "hi"}},
	})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if got := cap.header.Get("Authorization"); got != "Bearer "+testAPIKey {
		t.Errorf("Authorization = %q, want Bearer %s", got, testAPIKey)
	}
	if got := cap.header.Get("Content-Type"); got != "application/json" {
		t.Errorf("Content-Type = %q, want application/json", got)
	}
}

func TestBaseURLJoining(t *testing.T) {
	var cap captured
	client, srv := newTestClient(t, recordHandler(&cap, 200, `{"total":1}`))
	defer srv.Close()

	if _, err := client.Usage(context.Background()); err != nil {
		t.Fatalf("Usage: %v", err)
	}
	if cap.path != "/v1/usage" {
		t.Errorf("path = %q, want /v1/usage", cap.path)
	}
	if cap.method != http.MethodGet {
		t.Errorf("method = %q, want GET", cap.method)
	}
}

// -- each method hits the right path + verb -------------------------------------------

func TestChatCompletionsPathAndBody(t *testing.T) {
	reply := `{"id":"cmpl-1","choices":[{"message":{"content":"hi"}}]}`
	var cap captured
	client, srv := newTestClient(t, recordHandler(&cap, 200, reply))
	defer srv.Close()

	out, err := client.Chat.Completions.Create(context.Background(), ChatCompletionRequest{
		Model:    "smart",
		Messages: []Message{{Role: "user", Content: "hi"}},
		Extra:    map[string]any{"temperature": 0.2},
	})
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if string(out) != reply {
		t.Errorf("response = %s, want %s", out, reply)
	}
	if cap.path != "/v1/chat/completions" || cap.method != http.MethodPost {
		t.Errorf("got %s %s, want POST /v1/chat/completions", cap.method, cap.path)
	}

	var body map[string]any
	if err := json.Unmarshal(cap.body, &body); err != nil {
		t.Fatalf("decode body: %v", err)
	}
	if body["model"] != "smart" {
		t.Errorf("model = %v, want smart", body["model"])
	}
	if body["temperature"] != 0.2 {
		t.Errorf("temperature = %v, want 0.2", body["temperature"])
	}
	// Non-streaming requests explicitly send stream:false.
	if body["stream"] != false {
		t.Errorf("stream = %v, want false", body["stream"])
	}
}

func TestEmbeddingsStringInput(t *testing.T) {
	var cap captured
	client, srv := newTestClient(t, recordHandler(&cap, 200, `{"data":[{"embedding":[0.1]}]}`))
	defer srv.Close()

	if _, err := client.Embeddings.Create(context.Background(), EmbeddingRequest{
		Model: "embed",
		Input: "hello",
	}); err != nil {
		t.Fatalf("Create: %v", err)
	}
	if cap.path != "/v1/embeddings" {
		t.Errorf("path = %q, want /v1/embeddings", cap.path)
	}
	var body map[string]any
	_ = json.Unmarshal(cap.body, &body)
	if body["input"] != "hello" {
		t.Errorf("input = %v, want hello", body["input"])
	}
}

func TestEmbeddingsListInput(t *testing.T) {
	var cap captured
	client, srv := newTestClient(t, recordHandler(&cap, 200, `{"data":[]}`))
	defer srv.Close()

	if _, err := client.Embeddings.Create(context.Background(), EmbeddingRequest{
		Model: "embed",
		Input: []string{"a", "b"},
	}); err != nil {
		t.Fatalf("Create: %v", err)
	}
	var body struct {
		Input []string `json:"input"`
	}
	_ = json.Unmarshal(cap.body, &body)
	if len(body.Input) != 2 || body.Input[0] != "a" || body.Input[1] != "b" {
		t.Errorf("input = %v, want [a b]", body.Input)
	}
}

func TestImagesGenerate(t *testing.T) {
	var cap captured
	client, srv := newTestClient(t, recordHandler(&cap, 200, `{"data":[{"url":"http://img"}]}`))
	defer srv.Close()

	out, err := client.Images.Generate(context.Background(), ImageRequest{
		Model:  "image",
		Prompt: "a cat",
	})
	if err != nil {
		t.Fatalf("Generate: %v", err)
	}
	if cap.path != "/v1/images/generations" {
		t.Errorf("path = %q, want /v1/images/generations", cap.path)
	}
	var parsed struct {
		Data []struct {
			URL string `json:"url"`
		} `json:"data"`
	}
	if err := json.Unmarshal(out, &parsed); err != nil {
		t.Fatalf("decode response: %v", err)
	}
	if parsed.Data[0].URL != "http://img" {
		t.Errorf("url = %q, want http://img", parsed.Data[0].URL)
	}
	var body map[string]any
	_ = json.Unmarshal(cap.body, &body)
	if body["prompt"] != "a cat" {
		t.Errorf("prompt = %v, want a cat", body["prompt"])
	}
}

func TestModerations(t *testing.T) {
	var cap captured
	client, srv := newTestClient(t, recordHandler(&cap, 200,
		`{"flagged":true,"sensitiveData":{"email":1},"injection":[]}`))
	defer srv.Close()

	out, err := client.Moderations.Create(context.Background(), "email jane@acme.com")
	if err != nil {
		t.Fatalf("Create: %v", err)
	}
	if !out.Flagged {
		t.Error("Flagged = false, want true")
	}
	if out.SensitiveData["email"] != 1 {
		t.Errorf("SensitiveData[email] = %d, want 1", out.SensitiveData["email"])
	}
	if cap.path != "/v1/moderations" {
		t.Errorf("path = %q, want /v1/moderations", cap.path)
	}
	var body map[string]any
	_ = json.Unmarshal(cap.body, &body)
	if body["input"] != "email jane@acme.com" {
		t.Errorf("input = %v, want the email string", body["input"])
	}
}

func TestUsageModelsPolicies(t *testing.T) {
	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		switch r.URL.Path {
		case "/v1/usage":
			_, _ = io.WriteString(w, `{"total":5}`)
		case "/v1/models":
			_, _ = io.WriteString(w, `[{"alias":"smart","target":"gpt"}]`)
		case "/v1/policies":
			_, _ = io.WriteString(w, `[]`)
		default:
			w.WriteHeader(404)
		}
	})
	client, srv := newTestClient(t, handler)
	defer srv.Close()

	usage, err := client.Usage(context.Background())
	if err != nil {
		t.Fatalf("Usage: %v", err)
	}
	if string(usage) != `{"total":5}` {
		t.Errorf("usage = %s, want {\"total\":5}", usage)
	}

	models, err := client.Models(context.Background())
	if err != nil {
		t.Fatalf("Models: %v", err)
	}
	var list []map[string]any
	if err := json.Unmarshal(models, &list); err != nil {
		t.Fatalf("decode models: %v", err)
	}
	if len(list) != 1 || list[0]["alias"] != "smart" {
		t.Errorf("models = %s, want one entry aliased smart", models)
	}

	policies, err := client.Policies(context.Background())
	if err != nil {
		t.Fatalf("Policies: %v", err)
	}
	if string(policies) != `[]` {
		t.Errorf("policies = %s, want []", policies)
	}
}

func TestAuditQueryParams(t *testing.T) {
	var cap captured
	client, srv := newTestClient(t, recordHandler(&cap, 200, `{"entries":[],"total":0}`))
	defer srv.Close()

	if _, err := client.Audit(context.Background(), AuditParams{Q: "card", Verdict: "BLOCKED"}); err != nil {
		t.Fatalf("Audit: %v", err)
	}
	if cap.query.Get("q") != "card" {
		t.Errorf("q = %q, want card", cap.query.Get("q"))
	}
	if cap.query.Get("verdict") != "BLOCKED" {
		t.Errorf("verdict = %q, want BLOCKED", cap.query.Get("verdict"))
	}
}

func TestAuditOmitsEmptyParams(t *testing.T) {
	var cap captured
	client, srv := newTestClient(t, recordHandler(&cap, 200, `{"entries":[]}`))
	defer srv.Close()

	if _, err := client.Audit(context.Background(), AuditParams{}); err != nil {
		t.Fatalf("Audit: %v", err)
	}
	if cap.query.Has("q") {
		t.Error("q should be absent when not supplied")
	}
	if cap.query.Has("verdict") {
		t.Error("verdict should be absent when not supplied")
	}
}

// -- error envelope -> typed errors ---------------------------------------------------

func TestErrorEnvelopeMapsToSentinel(t *testing.T) {
	cases := []struct {
		status   int
		errType  string
		sentinel error
	}{
		{400, "invalid_request_error", ErrBadRequest},
		{401, "invalid_request_error", ErrAuthentication},
		{403, "policy_violation", ErrPolicyDenied},
		{403, "prompt_injection", ErrPolicyDenied},
		{404, "not_found", ErrNotFound},
		{429, "rate_limit_exceeded", ErrRateLimit},
		{502, "upstream_error", ErrUpstream},
		{503, "upstream_error", ErrUpstream},
		{504, "upstream_error", ErrUpstream},
	}

	for _, tc := range cases {
		tc := tc
		t.Run(http.StatusText(tc.status)+"_"+tc.errType, func(t *testing.T) {
			body := `{"error":{"message":"boom","type":"` + tc.errType + `","code":"x"}}`
			client, srv := newTestClient(t, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
				w.Header().Set("Content-Type", "application/json")
				w.WriteHeader(tc.status)
				_, _ = io.WriteString(w, body)
			}))
			defer srv.Close()

			_, err := client.Chat.Completions.Create(context.Background(), ChatCompletionRequest{
				Model:    "smart",
				Messages: []Message{{Role: "user", Content: "x"}},
			})
			if err == nil {
				t.Fatal("expected an error, got nil")
			}
			if !errors.Is(err, tc.sentinel) {
				t.Errorf("errors.Is(err, sentinel) = false; err = %v", err)
			}
			var apiErr *APIError
			if !errors.As(err, &apiErr) {
				t.Fatalf("errors.As(*APIError) = false; err = %v", err)
			}
			if apiErr.StatusCode != tc.status {
				t.Errorf("StatusCode = %d, want %d", apiErr.StatusCode, tc.status)
			}
			if apiErr.Message != "boom" {
				t.Errorf("Message = %q, want boom", apiErr.Message)
			}
			if apiErr.Type != tc.errType {
				t.Errorf("Type = %q, want %q", apiErr.Type, tc.errType)
			}
		})
	}
}

func TestErrorFallsBackToTypeForUnmappedStatus(t *testing.T) {
	err := errorFromResponse(418, map[string]any{
		"error": map[string]any{"message": "teapot", "type": "rate_limit_exceeded"},
	})
	if !errors.Is(err, ErrRateLimit) {
		t.Errorf("errors.Is(err, ErrRateLimit) = false; err = %v", err)
	}
	if err.Message != "teapot" {
		t.Errorf("Message = %q, want teapot", err.Message)
	}
}

func TestErrorHandlesNonObjectBody(t *testing.T) {
	err := errorFromResponse(500, "gateway exploded")
	if err.StatusCode != 500 {
		t.Errorf("StatusCode = %d, want 500", err.StatusCode)
	}
	if !strings.Contains(err.Message, "HTTP 500") {
		t.Errorf("Message = %q, want it to mention HTTP 500", err.Message)
	}
}

// -- transport errors -----------------------------------------------------------------

func TestConnectionErrorMapping(t *testing.T) {
	hc := &http.Client{Transport: roundTripperFunc(func(*http.Request) (*http.Response, error) {
		return nil, errors.New("dial tcp: connection refused")
	})}
	client, err := New(WithBaseURL("http://unreachable.test"), WithAPIKey(testAPIKey), WithHTTPClient(hc))
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	_, err = client.Usage(context.Background())
	if !errors.Is(err, ErrConnection) {
		t.Errorf("errors.Is(err, ErrConnection) = false; err = %v", err)
	}
}

func TestTimeoutErrorMapping(t *testing.T) {
	// The handler blocks until the request is cancelled, so the context deadline fires.
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		<-r.Context().Done()
	}))
	defer srv.Close()

	client, err := New(WithBaseURL(srv.URL), WithAPIKey(testAPIKey), WithHTTPClient(srv.Client()))
	if err != nil {
		t.Fatalf("New: %v", err)
	}
	ctx, cancel := context.WithTimeout(context.Background(), 20*time.Millisecond)
	defer cancel()
	_, err = client.Usage(ctx)
	if !errors.Is(err, ErrTimeout) {
		t.Errorf("errors.Is(err, ErrTimeout) = false; err = %v", err)
	}
}

File: sdk/js/test/client.test.ts
/**
 * Tests for the GateWise client. No network access — `fetch` is replaced with a stub on every
 * test, so each assertion can inspect the exact URL, verb, headers and body the client sent.
 */

import { afterEach, describe, expect, it, vi } from 'vitest';

import {
  GateWiseClient,
  AuthenticationError,
  BadRequestError,
  NotFoundError,
  PolicyDeniedError,
  RateLimitError,
  UpstreamError,
  GateWiseError,
  errorFromResponse,
} from '../src/index.js';

const BASE_URL = 'http://gateway.test';
const API_KEY = 'gatewise_sk_test_123';

/** Build a JSON `Response` like the real fetch would return. */
function jsonResponse(body: unknown, status = 200): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { 'Content-Type': 'application/json' },
  });
}

/** Build a fetch stub that records calls and returns a fixed (or computed) response. */
function stubFetch(
  responder: Response | ((url: string, init: RequestInit) => Response),
): { fetch: typeof fetch; calls: Array<{ url: string; init: RequestInit }> } {
  const calls: Array<{ url: string; init: RequestInit }> = [];
  const fetchImpl = vi.fn(async (url: string | URL | Request, init?: RequestInit) => {
    const u = String(url);
    const i = init ?? {};
    calls.push({ url: u, init: i });
    // Clone fixed responses so each call gets a readable body — a Response body is one-shot.
    return typeof responder === 'function' ? responder(u, i) : responder.clone();
  });
  return { fetch: fetchImpl as unknown as typeof fetch, calls };
}

function makeClient(responder: Response | ((url: string, init: RequestInit) => Response)) {
  const stub = stubFetch(responder);
  const client = new GateWiseClient({ baseUrl: BASE_URL, apiKey: API_KEY, fetch: stub.fetch });
  return { client, calls: stub.calls };
}

afterEach(() => {
  vi.restoreAllMocks();
});

// -- configuration / env ----------------------------------------------------------------

describe('configuration', () => {
  it('falls back to GATEWISE_* environment variables', () => {
    vi.stubEnv('GATEWISE_BASE_URL', 'http://env-host:9000');
    vi.stubEnv('GATEWISE_API_KEY', 'gatewise_sk_env');
    const client = new GateWiseClient({ fetch: stubFetch(jsonResponse({})).fetch });
    expect(client.baseUrl).toBe('http://env-host:9000');
    expect(client.apiKey).toBe('gatewise_sk_env');
    vi.unstubAllEnvs();
  });

  it('defaults the base URL to localhost:8080', () => {
    vi.stubEnv('GATEWISE_BASE_URL', '');
    const client = new GateWiseClient({ apiKey: 'k', fetch: stubFetch(jsonResponse({})).fetch });
    expect(client.baseUrl).toBe('http://localhost:8080');
    vi.unstubAllEnvs();
  });

  it('throws a clear error when no API key is configured', () => {
    vi.stubEnv('GATEWISE_API_KEY', '');
    expect(() => new GateWiseClient({ baseUrl: BASE_URL, fetch: stubFetch(jsonResponse({})).fetch }))
      .toThrow(/API key is required/);
    vi.unstubAllEnvs();
  });

  it('strips a trailing slash so the path is not doubled', async () => {
    const { client, calls } = (() => {
      const stub = stubFetch(jsonResponse({ total: 1 }));
      const c = new GateWiseClient({ baseUrl: BASE_URL + '/', apiKey: API_KEY, fetch: stub.fetch });
      return { client: c, calls: stub.calls };
    })();
    await client.usage();
    expect(calls[0]!.url).toBe(`${BASE_URL}/v1/usage`);
  });
});

// -- auth header + URL joining ----------------------------------------------------------

describe('request plumbing', () => {
  it('sends the Authorization: Bearer header on every request', async () => {
    const { client, calls } = makeClient(jsonResponse({ choices: [] }));
    await client.chat.completions.create({ model: 'smart', messages: [{ role: 'user', content: 'hi' }] });
    const headers = new Headers(calls[0]!.init.headers);
    expect(headers.get('Authorization')).toBe(`Bearer ${API_KEY}`);
    expect(headers.get('Content-Type')).toBe('application/json');
  });

  it('adds the /v1 prefix exactly once', async () => {
    const { client, calls } = makeClient(jsonResponse({ total: 1 }));
    await client.usage();
    expect(calls[0]!.url).toBe(`${BASE_URL}/v1/usage`);
    expect(calls[0]!.init.method).toBe('GET');
  });
});

// -- each method hits the right path + verb ---------------------------------------------

describe('methods', () => {
  it('chat.completions.create POSTs the right body', async () => {
    const reply = { id: 'cmpl-1', choices: [{ message: { content: 'hi' } }] };
    const { client, calls } = makeClient(jsonResponse(reply));
    const out = await client.chat.completions.create({
      model: 'smart',
      messages: [{ role: 'user', content: 'hi' }],
      temperature: 0.2,
    });
    expect(out).toEqual(reply);
    expect(calls[0]!.url).toBe(`${BASE_URL}/v1/chat/completions`);
    expect(calls[0]!.init.method).toBe('POST');
    const body = JSON.parse(calls[0]!.init.body as string);
    expect(body).toMatchObject({ model: 'smart', temperature: 0.2, stream: false });
  });

  it('embeddings.create accepts a string input', async () => {
    const { client, calls } = makeClient(jsonResponse({ data: [{ embedding: [0.1] }] }));
    await client.embeddings.create({ model: 'embed', input: 'hello' });
    expect(calls[0]!.url).toBe(`${BASE_URL}/v1/embeddings`);
    expect(JSON.parse(calls[0]!.init.body as string).input).toBe('hello');
  });

  it('embeddings.create accepts a list input', async () => {
    const { client, calls } = makeClient(jsonResponse({ data: [] }));
    await client.embeddings.create({ model: 'embed', input: ['a', 'b'] });
    expect(JSON.parse(calls[0]!.init.body as string).input).toEqual(['a', 'b']);
  });

  it('images.generate POSTs the prompt', async () => {
    const { client, calls } = makeClient(jsonResponse({ data: [{ url: 'http://img' }] }));
    const out = (await client.images.generate({ model: 'image', prompt: 'a cat' })) as {
      data: { url: string }[];
    };
    expect(out.data[0]!.url).toBe('http://img');
    expect(calls[0]!.url).toBe(`${BASE_URL}/v1/images/generations`);
    expect(JSON.parse(calls[0]!.init.body as string).prompt).toBe('a cat');
  });

  it('moderations.create POSTs input and returns the native result', async () => {
    const { client, calls } = makeClient(
      jsonResponse({ flagged: true, sensitiveData: { email: 1 }, injection: [] }),
    );
    const out = await client.moderations.create({ input: 'email jane@acme.com' });
    expect(out.flagged).toBe(true);
    expect(out.sensitiveData).toEqual({ email: 1 });
    expect(calls[0]!.url).toBe(`${BASE_URL}/v1/moderations`);
    expect(JSON.parse(calls[0]!.init.body as string)).toEqual({ input: 'email jane@acme.com' });
  });

  it('usage / models / policies GET the right paths', async () => {
    const { client, calls } = makeClient((url) => {
      if (url.endsWith('/usage')) return jsonResponse({ total: 5 });
      if (url.endsWith('/models')) return jsonResponse([{ alias: 'smart', target: 'gpt' }]);
      if (url.endsWith('/policies')) return jsonResponse([]);
      return jsonResponse({}, 404);
    });
    expect(await client.usage()).toEqual({ total: 5 });
    expect((await client.models()) as unknown[]).toHaveLength(1);
    expect(await client.policies()).toEqual([]);
    expect(calls.map((c) => c.init.method)).toEqual(['GET', 'GET', 'GET']);
  });

  it('audit serializes q/verdict and omits undefined params', async () => {
    const { client, calls } = makeClient(jsonResponse({ entries: [] }));
    await client.audit({ q: 'card', verdict: 'BLOCKED' });
    expect(calls[0]!.url).toBe(`${BASE_URL}/v1/audit?q=card&verdict=BLOCKED`);

    await client.audit();
    // No filters supplied -> no query string.
    expect(calls[1]!.url).toBe(`${BASE_URL}/v1/audit`);
  });
});

// -- error envelope -> typed exceptions -------------------------------------------------

describe('error mapping', () => {
  const cases: Array<[number, string, new (...a: never[]) => GateWiseError]> = [
    [400, 'invalid_request_error', BadRequestError],
    [401, 'invalid_request_error', AuthenticationError],
    [403, 'policy_violation', PolicyDeniedError],
    [403, 'prompt_injection', PolicyDeniedError],
    [404, 'not_found', NotFoundError],
    [429, 'rate_limit_exceeded', RateLimitError],
    [502, 'upstream_error', UpstreamError],
    [503, 'upstream_error', UpstreamError],
    [504, 'upstream_error', UpstreamError],
  ];

  for (const [status, type, Expected] of cases) {
    it(`maps HTTP ${status} (${type}) to ${Expected.name}`, async () => {
      const { client } = makeClient(
        jsonResponse({ error: { message: 'boom', type, code: 'x' } }, status),
      );
      await expect(
        client.chat.completions.create({ model: 'smart', messages: [{ role: 'user', content: 'x' }] }),
      ).rejects.toMatchObject({ message: 'boom', statusCode: status, type });
      // And it is an instance of the right class.
      const err = await client
        .chat.completions.create({ model: 'smart', messages: [{ role: 'user', content: 'x' }] })
        .catch((e) => e);
      expect(err).toBeInstanceOf(Expected);
      expect(err).toBeInstanceOf(GateWiseError);
    });
  }

  it('falls back to error.type when the status code is unmapped', () => {
    const err = errorFromResponse(418, { error: { message: 'teapot', type: 'rate_limit_exceeded' } });
    expect(err).toBeInstanceOf(RateLimitError);
    expect(err.message).toBe('teapot');
  });

  it('handles a non-object (plain text) error body', () => {
    const err = errorFromResponse(500, 'gateway exploded');
    expect(err.statusCode).toBe(500);
    expect(err.message).toContain('HTTP 500');
  });
});

// -- streaming --------------------------------------------------------------------------

/** Build a streaming Response whose body emits the given SSE text. */
function sseResponse(text: string, status = 200): Response {
  const stream = new ReadableStream<Uint8Array>({
    start(controller) {
      controller.enqueue(new TextEncoder().encode(text));
      controller.close();
    },
  });
  return new Response(stream, { status, headers: { 'Content-Type': 'text/event-stream' } });
}

describe('streaming', () => {
  it('yields each parsed SSE chunk and stops at [DONE]', async () => {
    const sse =
      'data: {"choices":[{"delta":{"content":"Hel"}}]}\n\n' +
      'data: {"choices":[{"delta":{"content":"lo"}}]}\n\n' +
      'data: [DONE]\n\n';
    const { client, calls } = makeClient(sseResponse(sse));

    const chunks: Record<string, unknown>[] = [];
    for await (const chunk of client.chat.completions.create({
      model: 'smart',
      messages: [{ role: 'user', content: 'hi' }],
      stream: true,
    })) {
      chunks.push(chunk);
    }

    expect(chunks).toHaveLength(2);
    expect((chunks[0]!.choices as any)[0].delta.content).toBe('Hel');
    expect((chunks[1]!.choices as any)[0].delta.content).toBe('lo');
    // The request advertised streaming to the gateway.
    expect(JSON.parse(calls[0]!.init.body as string).stream).toBe(true);
  });

  it('skips blank and malformed data lines without failing', async () => {
    const sse =
      '\n' +
      'data: \n\n' +
      'data: not-json\n\n' +
      'data: {"choices":[{"delta":{"content":"ok"}}]}\n\n' +
      'data: [DONE]\n\n';
    const { client } = makeClient(sseResponse(sse));

    const chunks: Record<string, unknown>[] = [];
    for await (const chunk of client.chat.completions.create({
      model: 'smart',
      messages: [{ role: 'user', content: 'hi' }],
      stream: true,
    })) {
      chunks.push(chunk);
    }
    expect(chunks).toHaveLength(1);
    expect((chunks[0]!.choices as any)[0].delta.content).toBe('ok');
  });

  it('raises a typed error on a non-2xx streaming response', async () => {
    const { client } = makeClient(
      jsonResponse({ error: { message: 'denied', type: 'policy_violation' } }, 403),
    );
    const iterate = async () => {
      for await (const _ of client.chat.completions.create({
        model: 'smart',
        messages: [{ role: 'user', content: 'hi' }],
        stream: true,
      })) {
        // no-op
      }
    };
    await expect(iterate()).rejects.toBeInstanceOf(PolicyDeniedError);
  });
});
